libexec/rtld-elf/aarch64/reloc.c

```diff
@@ -167,6 +177,224 @@ _rtld_relocate_nonplt_self(Elf_Dyn *dynp, Elf_Auxinfo *aux)
+
+	} else {
+
+		size_t new_len;
+		tramp_stk_table_t new_table;
+
+		new_len = len * 2;
+		new_table = xcalloc(new_len, sizeof(*new_table));
+		if (!new_table)
+			rtld_die();
+		assert(cheri_getlen(new_table) / sizeof(*new_table) == new_len);
+
+		for (size_t i = 0; i < len; ++i)
+			new_table[i] = table[i];
+
+		asm ("msr	ctpidr_el0, %0" :: "C" (new_table));
+		free(table);
+	}
+}
+
+static void (*thr_thread_start)(struct pthread *);
+
+void
+_rtld_thread_start(struct pthread *curthread)
+{
+	tramp_stk_table_t tls;
+	asm ("mrs	%0, ctpidr_el0" : "=C" (tls));
+	asm ("msr	rctpidr_el0, %0" :: "C" (tls));
+
+	tls = xcalloc(DEFAULT_STACK_TABLE_SIZE, sizeof(*tls));
+	tls[0] = (uintptr_t)get_rstk;
+	asm ("msr	ctpidr_el0, %0" :: "C" (tls));
+
+	thr_thread_start(curthread);
+}
+
+void
+_rtld_thread_start_init(void (*p)(struct pthread *))
+{
+	assert(!(cheri_getperm(p) & CHERI_PERM_EXECUTIVE));
+	assert(!thr_thread_start);
+	thr_thread_start = tramp_pgs_append(p, obj_from_addr(p));
+}
+
+static void (*thr_sighandler)(int, siginfo_t *, void *);
+
+void
+_rtld_sighandler(int sig, siginfo_t *info, void *_ucp)
+{
+	uintptr_t csp, rcsp;
+	ucontext_t *ucp = _ucp;
+
+	csp = ucp->uc_mcontext.mc_capregs.cap_sp;
+	asm ("mrs	%0, rcsp_el0" : "=C" (rcsp));
+	ucp->uc_mcontext.mc_capregs.cap_sp = rcsp;
+
+	thr_sighandler(sig, info, ucp);
+
+	rcsp = ucp->uc_mcontext.mc_capregs.cap_sp;
+	asm ("msr	rcsp_el0, %0" :: "C" (rcsp));
+	ucp->uc_mcontext.mc_capregs.cap_sp = csp;
+}
+
+void
+_rtld_sighandler_init(void *p)
+{
+	assert(!(cheri_getperm(p) & CHERI_PERM_EXECUTIVE));
+	assert(!thr_sighandler);
+	thr_sighandler = tramp_pgs_append(p, obj_from_addr(p));
+}
+
+static void *
+set_bounds_get_remainder(struct tramp **inout, size_t len)
+{
+	void *hi = *inout;
+	void *lo;
+	ptraddr_t top_lo, alignment;
+	size_t remaining;
+
+	*inout = lo = cheri_setbounds(hi, len);
+	top_lo = cheri_gettop(lo);
+	remaining = cheri_gettop(hi) - top_lo;
+	alignment = CHERI_REPRESENTABLE_ALIGNMENT(remaining);
+	return (cheri_setaddress(hi, roundup2(top_lo, alignment)));
+}
+
+static int
+tramp_pg_create(struct tramp_pg **out)
+{
+	struct tramp_pg *p;
+	p = mmap(NULL,
+		 getpagesize(),
+		 PROT_READ | PROT_WRITE | PROT_EXEC,
+		 MAP_ANON | MAP_PRIVATE,
+		 -1, 0);
+	if (p == MAP_FAILED)
+		return (-1);
+	p->cursor = p->trampolines;
+	*out = p;
+	return (0);
+}
+
+void *
+tramp_pgs_append(const void *target, const Obj_Entry *dst)
+{
+	extern const struct tramp tramp_template_exe, tramp_template_res;
+
+	static struct tramp_pgs pgs = SLIST_HEAD_INITIALIZER(pgs);
+
+	const struct tramp *template;
+	struct tramp *t;
+	size_t len;
+
+	int n_retry = 0;
+	struct tramp_pg *pg = SLIST_FIRST(&pgs);
+
+	if (!pg) {
+retry:
+		if (n_retry++)
+			rtld_die();
+		if (tramp_pg_create(&pg))
+			rtld_die();
+		SLIST_INSERT_HEAD(&pgs, pg, entries);
+	}
+
+	if (cheri_getperm(target) & CHERI_PERM_EXECUTIVE)
+		template = &tramp_template_exe;
+	else
+		template = &tramp_template_res;
+	len = cheri_getlen(template);
```

Comment:
Capability bounds are for catching errors not for being lazy and not keeping track of how big a variable is

Reply:
What is the recommended way for getting the size of such data? I see an example in `sigtramp.S` and wonder if that is it.

Reply:
Either that or exporting the end symbol and performing the subtraction at run time, yes